File: src/main/java/fi/quanfoxes/lexer/FunctionToken.java
package fi.quanfoxes.lexer;

import java.util.List;
import java.util.Objects;

import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Parser;

public class FunctionToken extends Token {
    private IdentifierToken name;
    private ContentToken parameters;

    /**
     * Creates a function token with name and parameters
     * @param name
     * @param parameters
     */
    public FunctionToken(IdentifierToken name, ContentToken parameters) {
        super(TokenType.FUNCTION);

        this.name = name;
        this.parameters = parameters;
    }

    /**
     * Returns the name of the function
     * @return Name of the function
     */
    public String getName() {
        return name.getValue();
    }

    /**
     * Parses the parameters with the given context
     * @param context Context used to parse parameters
     * @return Parameters in node tree form
     * @throws Exception Various reasons related to the parsing of the parameters
     */
    public Node getParameters(Context context) throws Exception {
        Node node = new Node();

        for (int i = 0; i < parameters.getSectionCount(); i++) {
            List<Token> tokens = parameters.getTokens(i);
            Parser.parse(node, context, tokens);
        }

        return node;
    }

    @Override
    public String getText() {
        return name.getText() + parameters.getText();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        FunctionToken that = (FunctionToken) o;
        return Objects.equals(name, that.name) &&
                Objects.equals(parameters, that.parameters);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, parameters);
    }
}

File: Zigzag/Lexer/FunctionToken.cs
using System;
using System.Collections.Generic;

public class FunctionToken : Token
{
	public IdentifierToken Identifier { get; private set; }
	public ContentToken Parameters { get; private set; }

	public string Name => Identifier.Value;

	public FunctionToken(IdentifierToken name, ContentToken parameters) : base(TokenType.FUNCTION)
	{
		Identifier = name;
		Parameters = parameters;
	}

	/// <summary>
	/// Returns function parameters as node tree
	/// </summary>
	/// <param name="context">Context used to parse</param>
	/// <returns>Parameters as node tree</returns>
	public Node GetParsedParameters(Context context)
	{
		var node = new Node();

		for (int i = 0; i < Parameters.SectionCount; i++)
		{
			var tokens = Parameters.GetTokens(i);
			Parser.Parse(node, context, tokens);
		}

		return node;
	}

	/// <summary>
	/// Returns the parameter names
	/// </summary>
	/// <returns>List of parameter names</returns>
	public List<string> GetParameterNames()
	{
		var names = new List<string>();

		if (Parameters.IsEmpty)
		{
			return names;
		}

		for (int i = 0; i < Parameters.SectionCount; i++)
		{
			var tokens = Parameters.GetTokens(i);

			if (tokens.Count > 1)
			{
				throw Errors.Get(tokens[0].Position, "Advanced parameters aren't supported yet!");
			}

			var token = tokens[0];

			if (!(token is IdentifierToken name))
			{
				throw Errors.Get(token.Position, "Invalid parameter");
			}

			names.Add(name.Value);
		}

		return names;
	}

	public override bool Equals(object obj)
	{
		return obj is FunctionToken token &&
			   base.Equals(obj) &&
			   EqualityComparer<IdentifierToken>.Default.Equals(Identifier, token.Identifier) &&
			   EqualityComparer<ContentToken>.Default.Equals(Parameters, token.Parameters) &&
			   Name == token.Name;
	}

	public override int GetHashCode()
	{
		return HashCode.Combine(base.GetHashCode(), Identifier, Parameters, Name);
	}
}
